File: data/TiendaMusical/TiendaMusica.java
import java.util.ArrayList;
import java.util.HashMap;

/**
 * Clase que gestiona la tienda de música.
 */
public class TiendaMusica {
    private String nombre;
    private ArrayList<Producto> inventario;
    private HashMap<String, Cliente> clientes;

    public TiendaMusica(String nombre) {
        this.nombre = nombre;
        this.inventario = new ArrayList<>();
        this.clientes = new HashMap<>();
    }

    /**
     * Añade un producto al inventario.
     * Debe validar que no exista un producto con el mismo ID antes de añadirlo.
     * Si el producto ya existe, no se añade
     * @param producto Producto a añadir
     */
    public void añadirProducto(Producto producto) {
        //TO DO
    }

    /**
     * Añade un cliente a la tienda.
     * @param cliente Cliente a añadir
     */

    public void añadirCliente(Cliente cliente) {
        clientes.put(cliente.getId(), cliente);
    }


    /**
     * Registra una nueva venta.
     * Solo debe agregar la venta al historial del cliente.
     * @param clienteId ID del cliente
     * @param productoId ID del producto
     * @param cantidad Cantidad vendida
     */
    public void registrarVenta(String clienteId, String productoId, int cantidad) {
        // TODO: Implementar lógica de venta
    }

    // Setters y Getters
    public String getNombre() {
        return nombre;
    }
    public ArrayList<Producto> getInventario() {
        return new ArrayList<>(inventario);
    }
    public HashMap<String, Cliente> getClientes() {
        return new HashMap<>(clientes);
    }
}

File: data/TiendaMusical/Dependencies/TiendaMusica.js
/**
 * Clase que gestiona la tienda de música.
 */
class TiendaMusica {
    /**
     * @param {string} nombre 
     */
    constructor(nombre) {
        this.nombre = nombre;
        /** @type {Producto[]} */
        this.inventario = [];
        /** @type {Object.<string, Cliente>} */
        this.clientes = {};
    }

    /**
     * Añade un producto al inventario.
     * Valida que no exista un producto con el mismo ID antes de añadirlo.
     * @param {Producto} producto Producto a añadir
     */
    añadirProducto(producto) {
        if (producto && !this.inventario.some(p => p.getId() === producto.getId())) {
            this.inventario.push(producto);
        }
    }

    /**
     * Añade un cliente a la tienda.
     * @param {Cliente} cliente Cliente a añadir
     */
    añadirCliente(cliente) {
        if (cliente) {
            this.clientes[cliente.getId()] = cliente;
        }
    }

    /**
     * Registra una nueva venta.
     * Agrega la venta al historial del cliente.
     * @param {string} clienteId ID del cliente
     * @param {string} productoId ID del producto
     * @param {number} cantidad Cantidad vendida
     * @throws {Error} Si el cliente o producto no existen o la cantidad es inválida
     */
    registrarVenta(clienteId, productoId, cantidad) {
        if (!this.clientes[clienteId]) {
            throw new Error("Cliente no encontrado");
        }
        
        if (!this.inventario.some(p => p.getId() === productoId)) {
            throw new Error("Producto no encontrado");
        }
        
        if (cantidad <= 0) {
            throw new Error("La cantidad debe ser positiva");
        }
        
        this.clientes[clienteId].añadirCompra(productoId, cantidad);
    }

    // Getters
    getNombre() {
        return this.nombre;
    }
    
    getInventario() {
        return [...this.inventario]; // Devuelve copia del array
    }
    
    getClientes() {
        return {...this.clientes}; // Devuelve copia del objeto
    }
}